### DISCUSSION TITLE
Is it possible to write shared tests?

### DISCUSSION THREAD

**Discussion by mertasan** [pestphp/pest#335] [Q&A]
My test file looks like the example below.

```php
<?php

test('foo 1', function () {
    
    $this->assertCount(myTestHelper()->results()->count(), $this->dummyData);

});

test('foo 2', function () {
    //...
});

test('bar 1', function () {
    //...
});

test('bar 2', function () {
    //...
});

beforeEach(function () {
    
    $this->dummy = [
        'name'   => 'user name %s',
    ];
    
    $this->totalRow = 2;
    $this->dummyData = dummyData($this->dummy, $this->totalRow);
    
    migrate();
    
    foreach ($this->dummyData as $index => $data) {
        
        // factories
        user([
            'name'       => $data['name'], // user name 1, user name 2
        ]);
    }
    
    $this->myProperty = myHelper(MyTestComponent::class, debug: false);
    
});

beforeAll(function () {
    
    class MyTestComponent extends TableComponent
    {
        
        public function columns (): array
        {
            return [
                
                Column::string('name'),
                // ...
                // ...
                // ...
            ];
        }
    }
});
```

Is it possible to share some tests with other test files?

Example:

($this->dummyData should take different data in each file. or some other way i need to be able to access this data.)

```php
// Pest.php
beforeEach(function () {
    
    test('foo test', function () {
        
        $this->assertCount(myTestHelper()->results()->count(), $this->dummyData);
        
    })->group('group-foo');
    
    test('bar test', function () {
        
        $this->assertCount(myTestHelper()->results()->count(), $this->dummyData);
        
    })->group('group-bar');
    
});
```

or `groupEach`, `groupAll` etc:

```php
// Pest.php
groupAll('group-foo', function () {

    test('foo test', function () {
        
        $this->assertCount(myTestHelper()->results()->count(), $this->dummyData);
        
    });

});

groupAll('group-bar', function () {
    
    test('bar test', function () {
        
        $this->assertCount(myTestHelper()->results()->count(), $this->dummyData);
        
    });
    
});
```

```php
// Foo1Test.php
uses()->group('group-foo');
```
```php
// Foo2Test.php
uses()->group('group-foo');
```
```php
// Bar1Test.php
uses()->group('group-bar');
```
```php
// Bar2Test.php
uses()->group('group-bar');
```

Thanks to everyone who contributed to this amazing package!

**Edit:**

or something like:

```php
// TestGroupFoo.php
uses()->group('group-foo');

test('test foo', function(){
    $this->assertCount(myTestHelper()->results()->count(), $this->dummyData);
});
```

```php
// TestGroupBar.php
uses()->group('group-bar');

test('test bar', function(){
    $this->assertSame(myTestHelper()->results(), $this->dummyData);
});
```

```php
// Foo1Test.php
uses()->merge('group-foo', 'group-bar');
```

```php
// Bar1Test.php
uses()->merge('group-foo');
```

```php
// Foo1Test.php and Bar1Test.php (Both files will have different data.)

beforeEach(function () {
    
    $this->dummy = [
        'name'   => 'user name %s',
    ];
    
    $this->totalRow = 2;
    $this->dummyData = dummyData($this->dummy, $this->totalRow);
    
    migrate();
    
    foreach ($this->dummyData as $index => $data) {
        
        // factories
        user([
            'name'       => $data['name'], // user name 1, user name 2
        ]);
    }
    
    $this->myProperty = myHelper(MyTestComponent::class, debug: false);
    
});

beforeAll(function () {
    
    class MyTestComponent extends TableComponent
    {
        
        public function columns (): array
        {
            return [
                
                Column::string('name'),
                // ...
                // ...
                // ...
            ];
        }
    }
});
```

**Comment by nelson6e65** [comment]
I used to shared tests in PHPUnit via traits, sharing data providers and generic test implementations using traits:

- Specific TestCase using a trait as test helper: [VersionTest.php#L37](https://github.com/nelson6e65/php_nml/blob/3c4141215981577ccc7da544c4cc2bebf9317ff8/tests/TestCase/VersionTest.php#L37).
- Trait with providers for that specific test case using a helper: [VersionTestProvider.php#L51](https://github.com/nelson6e65/php_nml/blob/3c4141215981577ccc7da544c4cc2bebf9317ff8/tests/DataProviders/VersionTestProvider.php#L51) (`ImplementsIConvertibleToString` in this example).
- Generic test's helper used in that trait, which only changes depending on the data-provider defined (abstract methods): 
  - Data provider definition: [ImplementsIConvertibleToString#L44](https://github.com/nelson6e65/php_nml/blob/3c4141215981577ccc7da544c4cc2bebf9317ff8/tests/Helpers/ImplementsIConvertibleToString.php#L44)
  - Usage in the helper trait: [ImplementsIConvertibleToString.php#L68](https://github.com/nelson6e65/php_nml/blob/3c4141215981577ccc7da544c4cc2bebf9317ff8/tests/Helpers/ImplementsIConvertibleToString.php#L68)
  - Data provider implementation: [VersionTestProvider.php#L334](https://github.com/nelson6e65/php_nml/blob/3c4141215981577ccc7da544c4cc2bebf9317ff8/tests/DataProviders/VersionTestProvider.php#L334)

So, if I need to use that trait helper (`ImplementsIConvertibleToString`), or any other, I only have to implement the data provider method (`toStringProvider()`) in any class, by using the same trait, only changing the provider. For example:[VersionComponentTestProvider.php#L243](https://github.com/nelson6e65/php_nml/blob/3c4141215981577ccc7da544c4cc2bebf9317ff8/tests/DataProviders/VersionComponentTestProvider.php#L243) and [TypeTestProvider.php#L104](https://github.com/nelson6e65/php_nml/blob/3c4141215981577ccc7da544c4cc2bebf9317ff8/tests/DataProviders/TypeTestProvider.php#L104).

As you can see, you can avoid a lot of code duplication in your tests with this approach.

I think that is what you, and I, want, but in Pest PHP. I'll check how to do this in Pest.

 I'll be in touch. 😊

**Comment by SDIjeremy** [comment]
Any solution to this? Seems a shortcoming of Pest where we can't use it like a class and extend or add traits?
